File: src-tauri/src/hooks.rs
use std::path::PathBuf;

/// 执行所有 trigger 对应的脚本。
/// 配置格式（在 settings.json 中）：
/// ```json
/// { "hooks": { "startup": ["./hooks/fetch-key.sh"] } }
/// ```
/// 脚本 stdout 按行解析 `KEY=VALUE` 并写入环境变量。
pub fn run_hooks(trigger: &str) {
    let base = match pagesprite_dir() {
        Ok(d) => d,
        Err(e) => {
            eprintln!("[hooks] {e}");
            return;
        }
    };

    let scripts = match load_scripts(trigger, &base) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("[hooks] {e}");
            return;
        }
    };

    for script_rel in &scripts {
        let script_path = base.join(script_rel);
        if !script_path.exists() {
            eprintln!("[hooks] script not found: {}", script_path.display());
            continue;
        }

        eprintln!("[hooks] running: {}", script_rel);
        match exec_script(&script_path) {
            Ok(lines) => {
                for (key, val) in &lines {
                    std::env::set_var(key, val);
                    eprintln!("[hooks]   ${} = {} chars", key, val.len());
                }
                if lines.is_empty() {
                    eprintln!("[hooks]   (no output)");
                }
            }
            Err(e) => {
                eprintln!("[hooks] {script_rel} failed: {e}");
            }
        }
    }
}

/// 兼容旧的 run_startup_hooks 调用方式
pub fn run_startup_hooks() {
    run_hooks("startup");
}

/// 从 settings.json 读取 hooks 配置，返回指定 trigger 的脚本路径列表。
fn load_scripts(trigger: &str, base: &PathBuf) -> Result<Vec<String>, String> {
    let path = base.join("settings.json");
    if !path.exists() {
        return Ok(Vec::new());
    }
    let content = std::fs::read_to_string(&path)
        .map_err(|e| format!("cannot read {}: {e}", path.display()))?;
    let raw: serde_json::Value =
        serde_json::from_str(&content).map_err(|e| format!("parse error: {e}"))?;

    let Some(arr) = raw
        .get("hooks")
        .and_then(|v| v.get(trigger))
        .and_then(|v| v.as_array())
    else {
        return Ok(Vec::new());
    };

    let scripts: Vec<String> = arr
        .iter()
        .filter_map(|v| v.as_str().map(String::from))
        .collect();
    Ok(scripts)
}

/// 执行脚本，解析 stdout 中的 KEY=VALUE 行。
fn exec_script(script_path: &PathBuf) -> Result<Vec<(String, String)>, String> {
    let ext = script_path
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_string();

    let (program, args): (&str, Vec<&str>) = if cfg!(target_os = "windows") {
        match ext.as_str() {
            "bat" | "cmd" => ("cmd", vec!["/c", script_path.to_str().unwrap()]),
            "js" => ("node", vec![script_path.to_str().unwrap()]),
            "py" => ("python", vec![script_path.to_str().unwrap()]),
            "ps1" => ("powershell", vec!["-ExecutionPolicy", "Bypass", "-File", script_path.to_str().unwrap()]),
            _ => return Err(format!("unsupported script type on Windows: .{ext}")),
        }
    } else {
        match ext.as_str() {
            "sh" => ("sh", vec![script_path.to_str().unwrap()]),
            "js" => ("node", vec![script_path.to_str().unwrap()]),
            "py" => ("python3", vec![script_path.to_str().unwrap()]),
            _ => return Err(format!("unsupported script type: .{ext}")),
        }
    };

    let output = std::process::Command::new(program)
        .args(&args)
        .output()
        .map_err(|e| format!("cannot spawn: {e}"))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!(
            "exit {:?}: {}",
            output.status.code(),
            stderr.trim()
        ));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let mut result = Vec::new();
    for line in stdout.lines() {
        let line = line.trim();
        if let Some(eq_pos) = line.find('=') {
            let key = line[..eq_pos].trim().to_string();
            let val = line[eq_pos + 1..].trim().to_string();
            if !key.is_empty() {
                result.push((key, val));
            }
        }
    }
    Ok(result)
}

fn pagesprite_dir() -> Result<PathBuf, String> {
    let home = dirs::home_dir().ok_or("cannot find home directory")?;
    Ok(home.join(".pagesprite"))
}

File: electron/ipc/hooks.js
import fs from "fs";
import path from "path";
import { homedir } from "os";
import { spawnSync } from "child_process";

const BASE_DIR = path.join(homedir(), ".pagesprite");
const SETTINGS_PATH = path.join(BASE_DIR, "settings.json");

/** Load hooks config from settings.json */
function loadScripts(trigger) {
  try {
    if (!fs.existsSync(SETTINGS_PATH)) return [];
    const raw = JSON.parse(fs.readFileSync(SETTINGS_PATH, "utf-8"));
    const scripts = raw.hooks?.[trigger];
    return Array.isArray(scripts) ? scripts.filter((s) => typeof s === "string") : [];
  } catch {
    return [];
  }
}

/** Run a single script and parse KEY=VALUE lines from stdout */
function execScript(scriptPath) {
  if (!fs.existsSync(scriptPath)) {
    console.error(`[hooks] script not found: ${scriptPath}`);
    return [];
  }

  const ext = path.extname(scriptPath).slice(1).toLowerCase();
  let program, args;

  if (process.platform === "win32") {
    switch (ext) {
      case "bat":
      case "cmd":
        program = "cmd";
        args = ["/c", scriptPath];
        break;
      case "js":
        program = "node";
        args = [scriptPath];
        break;
      case "py":
        program = "python";
        args = [scriptPath];
        break;
      case "ps1":
        program = "powershell";
        args = ["-ExecutionPolicy", "Bypass", "-File", scriptPath];
        break;
      default:
        console.error(`[hooks] unsupported script type on Windows: .${ext}`);
        return [];
    }
  } else {
    switch (ext) {
      case "sh":
        program = "sh";
        args = [scriptPath];
        break;
      case "js":
        program = "node";
        args = [scriptPath];
        break;
      case "py":
        program = "python3";
        args = [scriptPath];
        break;
      default:
        console.error(`[hooks] unsupported script type: .${ext}`);
        return [];
    }
  }

  console.error(`[hooks] running: ${path.basename(scriptPath)}`);
  const result = spawnSync(program, args, { timeout: 30000 });

  if (result.error) {
    console.error(`[hooks] failed: ${result.error.message}`);
    return [];
  }
  if (result.status !== 0) {
    const stderr = result.stderr?.toString().trim();
    console.error(`[hooks] exit ${result.status}: ${stderr}`);
    return [];
  }

  const stdout = result.stdout?.toString() || "";
  const lines = [];
  for (const line of stdout.split("\n")) {
    const trimmed = line.trim();
    const eqIdx = trimmed.indexOf("=");
    if (eqIdx > 0) {
      const key = trimmed.slice(0, eqIdx).trim();
      const val = trimmed.slice(eqIdx + 1).trim();
      if (key) {
        process.env[key] = val;
        lines.push({ key, val });
        console.error(`[hooks]   $${key} = ${val.length} chars`);
      }
    }
  }
  if (lines.length === 0) console.error("[hooks]   (no output)");
  return lines;
}

/** Run all scripts for a given trigger */
export function runHooks(trigger) {
  const scripts = loadScripts(trigger);
  for (const rel of scripts) {
    execScript(path.join(BASE_DIR, rel));
  }
}

/** Run startup hooks (convenience wrapper) */
export function runStartupHooks() {
  runHooks("startup");
}
